### PR TITLE
Ansible module for storage upload/download

### PR DESCRIPTION
**Pull request by rambleraptor**
Lots of custom code in this module! I took an approach of building a new "template" specifically for this module.

This new template uses the building blocks of the regular Ansible modules (so we get autoupdating documentation, arg parsing for free), but ignores all of the other stuff.

This PR contains:
* Moving documentation into a new file so it can be easily reused by the object upload/download module
* API.yaml info for the module
* Support for specifying which template should be used
* A object_storage template with a bunch of custom code.

This module currently uploads/downloads a file. It doesn't handle overwrite logic. That'll come next.

-----------------------------------------------------------------
# [all]

### CODE REVIEW

File: products/storage/examples/ansible/object.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+--- !ruby/object:Provider::Ansible::Example
+task: !ruby/object:Provider::Ansible::Task
+  name: gcp_storage_object
+  code:
+    name: 'ansible-storage-module'
+    action: download
+    bucket: ansible-bucket
+    src: modules.zip
+    dest: ~/modules.zip
+    project: google.com:graphite-playground
```
Comment (SirGitsalot) [inline products/storage/examples/ansible/object.yaml]:
```suggestion
    project: <%= ctx[:project] %>
```

File: products/storage/examples/ansible/object.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    auth_kind: serviceaccount
```
Comment (SirGitsalot) [inline products/storage/examples/ansible/object.yaml]:
```suggestion
    auth_kind: <%= ctx[:auth_kind] %>
```

File: products/storage/examples/ansible/object.yaml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    service_account_file: /Users/alexstephen/serviceaccounts/graphite_playground.json
```
Comment (SirGitsalot) [inline products/storage/examples/ansible/object.yaml]:
```suggestion
    service_account_file: <%= ctx[:service_account_file] %>
```

File: products/storage/api.yaml [hunk 4]
```diff
@@ -21,6 +21,35 @@ versions:
...
+    description: Upload or download a file from a GCS bucket.
+    exclude: true
+    parameters:
+      - !ruby/object:Api::Type::Enum
+        name: 'action'
+        description: 'Upload or download from the bucket'
+        values:
+          - :download
+          - :upload
+      # TODO(alexstephen): Figure out why MD5 hashes are different local vs. remote
```
Comment (SirGitsalot) [inline products/storage/api.yaml]: That sounds bad

File: products/storage/helpers/ansible/object_template.erb [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        # Get contents of file
+        results = download_file(module)
+
+    module.exit_json(**results)
+
+
+def upload_file(module):
+    auth = GcpSession(module, 'storage')
+    data = auth.get(media_link(module))
+    f = open(module.params['dest'], 'w')
```
Comment (SirGitsalot) [inline products/storage/helpers/ansible/object_template.erb]: `with open(module.params['dest'], 'w') as f:`

File: products/storage/helpers/ansible/object_template.erb [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    f.write(data.text.encode('utf8'))
+    f.close()
+    return fetch_resource(module, self_link(module))
+
+
+def download_file(module):
+    auth = GcpSession(module, 'storage')
+    f = open(module.params['src'], 'r')
```
Comment (SirGitsalot) [inline products/storage/helpers/ansible/object_template.erb]: Ditto on the `with`

File: products/storage/helpers/ansible/object_template.erb [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+    if not module.params['scopes']:
+        module.params['scopes'] = <%= python_literal(object.__product.scopes) %>
+
+    results = module.params
+
+    auth = GcpSession(module, 'storage')
+    if module.params['action'] == 'download':
+        # Get file from API
+        results = upload_file(module)
```
Comment (SirGitsalot) [inline products/storage/helpers/ansible/object_template.erb]:
```
if module.params['action'] == 'download':
        # Get file from API
        results = upload_file(module)
```
If download then upload?

File: products/storage/api.yaml (hunk 4 above)
Comment (rambleraptor) [reply products/storage/api.yaml]: Yeah, that stumped me for the longest time. Turns out that I was doing everything the hard way. The `requests` library handles all of the file-upload logic for me (and it does it correctly!)

**Pull request merged**